File: backend/app/api/requests.py
from fastapi import APIRouter, HTTPException, status, Depends, Query
from pydantic import BaseModel
from typing import Optional, List, Dict, Any
from datetime import datetime
from app.services.request_service import RequestService
from app.core.security import get_current_user, get_current_user_id
import logging

logger = logging.getLogger(__name__)

router = APIRouter()
request_service = RequestService()


# Request/Response Models
class BloodRequestCreate(BaseModel):
    patientName: str
    requiredBloodGroup: str
    city: str
    gender: Optional[str] = None
    hospital: Optional[str] = None
    locationAddress: Optional[str] = None
    locationLat: Optional[float] = None
    locationLng: Optional[float] = None
    unitsRequired: Optional[int] = None
    neededBy: Optional[datetime] = None
    notes: Optional[str] = None
    requestedTo: Optional[str] = None


@router.post("/")
async def create_blood_request(
    request_data: BloodRequestCreate,
    current_user: Dict[str, Any] = Depends(get_current_user)
):
    """Create a new blood request."""
    try:
        user_id = current_user["user_id"]
        
        request = await request_service.create_blood_request(
            created_by=user_id,
            patient_name=request_data.patientName,
            required_blood_group=request_data.requiredBloodGroup,
            city=request_data.city,
            gender=request_data.gender,
            hospital=request_data.hospital,
            locationAddress=request_data.locationAddress,
            locationLat=request_data.locationLat,
            locationLng=request_data.locationLng,
            unitsRequired=request_data.unitsRequired,
            neededBy=request_data.neededBy,
            notes=request_data.notes,
            requestedTo=request_data.requestedTo
        )
        
        return {
            "success": True,
            "data": request.__dict__,
            "message": "Blood request created successfully"
        }
        
    except Exception as e:
        logger.error(f"Create blood request error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to create blood request"
        )


@router.get("/")
async def list_requests(
    status: Optional[str] = Query(None, description="Filter by status"),
    city: Optional[str] = Query(None, description="Filter by city"),
    requiredBloodGroup: Optional[str] = Query(None, description="Filter by blood group"),
    mineOnly: bool = Query(False, description="Show only my requests"),
    toMeOnly: bool = Query(False, description="Show only requests to me"),
    openOnly: bool = Query(False, description="Show only open requests"),
    includeMatchScores: bool = Query(False, description="Include match scores for current user"),
    skip: int = Query(0, ge=0, description="Number of records to skip"),
    limit: int = Query(100, ge=1, le=1000, description="Number of records to return"),
    current_user: Dict[str, Any] = Depends(get_current_user)
):
    """List blood requests with filters. Can include match scores if user is a donor."""
    try:
        user_id = current_user["user_id"]
        
        filters = {}
        if status:
            filters["status"] = status
        if city:
            filters["city"] = city
        if requiredBloodGroup:
            filters["requiredBloodGroup"] = requiredBloodGroup
        if mineOnly:
            filters["mineOnly"] = True
            filters["user_id"] = user_id
        if toMeOnly:
            filters["toMeOnly"] = True
            filters["user_id"] = user_id
        if openOnly:
            filters["openOnly"] = True
        
        requests = await request_service.list_requests(
            filters, skip, limit,
            include_match_scores=includeMatchScores,
            viewer_user_id=user_id if includeMatchScores else None
        )
        
        # Convert to dictionaries and include match scores if present
        request_dicts = []
        for req in requests:
            req_dict = req.__dict__.copy()
            # Include match score data if present
            if hasattr(req, 'match_score_data'):
                req_dict['matchScore'] = req.match_score_data
            request_dicts.append(req_dict)
        
        return {
            "success": True,
            "data": request_dicts,
            "pagination": {
                "skip": skip,
                "limit": limit,
                "count": len(request_dicts)
            }
        }
        
    except Exception as e:
        logger.error(f"List requests error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to list requests"
        )


@router.get("/inbox")
async def get_donor_inbox(current_user: Dict[str, Any] = Depends(get_current_user)):
    """Get donor inbox (targeted + discoverable requests)."""
    try:
        user_id = current_user["user_id"]
        
        inbox = await request_service.get_donor_inbox(user_id)
        
        return {
            "success": True,
            "data": {
                "targeted": [req.__dict__ for req in inbox["targeted"]],
                "discoverable": [req.__dict__ for req in inbox["discoverable"]]
            }
        }
        
    except Exception as e:
        logger.error(f"Get donor inbox error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to get donor inbox"
        )


@router.get("/{request_id}")
async def get_request_by_id(
    request_id: str,
    current_user: Dict[str, Any] = Depends(get_current_user)
):
    """Get specific blood request."""
    try:
        request = await request_service.get_request_by_id(request_id)
        
        if not request:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Request not found"
            )
        
        return {
            "success": True,
            "data": request.__dict__
        }
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Get request by ID error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to get request"
        )


@router.put("/{request_id}/accept")
async def accept_request(
    request_id: str,
    current_user: Dict[str, Any] = Depends(get_current_user)
):
    """Accept a blood request (donor action)."""
    try:
        user_id = current_user["user_id"]
        
        updated_request = await request_service.accept_request(request_id, user_id)
        
        if not updated_request:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Request not found"
            )
        
        return {
            "success": True,
            "data": updated_request.__dict__,
            "message": "Request accepted successfully"
        }
        
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Accept request error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to accept request"
        )


@router.put("/{request_id}/reject")
async def reject_request(
    request_id: str,
    current_user: Dict[str, Any] = Depends(get_current_user)
):
    """Reject a blood request (donor action)."""
    try:
        user_id = current_user["user_id"]
        
        updated_request = await request_service.reject_request(request_id, user_id)
        
        if not updated_request:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Request not found"
            )
        
        return {
            "success": True,
            "data": updated_request.__dict__,
            "message": "Request rejected"
        }
        
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Reject request error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to reject request"
        )


@router.put("/{request_id}/cancel")
async def cancel_request(
    request_id: str,
    current_user: Dict[str, Any] = Depends(get_current_user)
):
    """Cancel a blood request (creator action)."""
    try:
        user_id = current_user["user_id"]
        
        updated_request = await request_service.cancel_request(request_id, user_id)
        
        if not updated_request:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Request not found"
            )
        
        return {
            "success": True,
            "data": updated_request.__dict__,
            "message": "Request cancelled"
        }
        
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Cancel request error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to cancel request"
        )


@router.put("/{request_id}/fulfill")
async def mark_request_fulfilled(
    request_id: str,
    current_user: Dict[str, Any] = Depends(get_current_user)
):
    """Mark a blood request as fulfilled (creator action)."""
    try:
        user_id = current_user["user_id"]
        
        updated_request = await request_service.mark_request_fulfilled(request_id, user_id)
        
        if not updated_request:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Request not found"
            )
        
        return {
            "success": True,
            "data": updated_request.__dict__,
            "message": "Request marked as fulfilled"
        }
        
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e)
        )
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Mark request fulfilled error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to mark request as fulfilled"
        )

File: mobile-app/lib/requests.ts
import { auth, database } from '@/database/firebase';
import { get, push, query, ref, set } from 'firebase/database';
import { createDonationRecord } from './donations';
import { BloodRequest } from './types';

function stripUndefined<T extends Record<string, unknown>>(obj: T): T {
  const cleaned = Object.fromEntries(
    Object.entries(obj).filter(([, v]) => v !== undefined)
  ) as T;
  return cleaned;
}

export async function postRequest(payload: Omit<BloodRequest, 'id' | 'status' | 'createdAt' | 'createdBy'>): Promise<string> {
  const uid = auth.currentUser?.uid;
  if (!uid) throw new Error('Not authenticated');
  const id = push(ref(database, 'requests')).key!;
  const data: BloodRequest = {
    id,
    createdBy: uid,
    status: payload.requestedTo ? 'pending' : 'open',
    createdAt: Date.now(),
    patientName: payload.patientName,
    requiredBloodGroup: payload.requiredBloodGroup,
    city: payload.city,
    gender: payload.gender,
    hospital: payload.hospital,
    locationAddress: (payload as any).locationAddress,
    locationLat: (payload as any).locationLat,
    locationLng: (payload as any).locationLng,
    unitsRequired: payload.unitsRequired,
    neededBy: payload.neededBy,
    notes: payload.notes,
    requestedTo: payload.requestedTo,
    urgent: payload.urgent,
  };
  await set(ref(database, `requests/${id}`), stripUndefined(data));
  return id;
}

export async function listMyRequests(uid?: string): Promise<BloodRequest[]> {
  const userId = uid ?? auth.currentUser?.uid;
  if (!userId) return [];
  const snap = await get(ref(database, 'requests'));
  if (!snap.exists()) return [];
  const all: BloodRequest[] = Object.values(snap.val());
  return all.filter((r) => r.createdBy === userId);
}

export async function getRequestById(id: string): Promise<BloodRequest | null> {
  const snap = await get(ref(database, `requests/${id}`));
  if (!snap.exists()) return null;
  return snap.val() as BloodRequest;
}

export async function getDonorStats(uid: string): Promise<{ totalReceived: number; accepted: number; rejected: number; }>{
  const snap = await get(ref(database, 'requests'));
  if (!snap.exists()) return { totalReceived: 0, accepted: 0, rejected: 0 };
  const all: BloodRequest[] = Object.values(snap.val());
  const mine = all.filter((r) => r.requestedTo === uid);
  const accepted = mine.filter((r) => r.status === 'accepted').length;
  const rejected = mine.filter((r) => r.status === 'rejected').length;
  return { totalReceived: mine.length, accepted, rejected };
}

export async function canAcceptRequest(requestId: string, donorUid?: string): Promise<boolean> {
  const uid = donorUid ?? auth.currentUser?.uid;
  if (!uid) return false;
  
  try {
    // Check if donor is available
    const userSnap = await get(ref(database, `users/${uid}`));
    if (!userSnap.exists()) return false;
    const userProfile = userSnap.val();
    if (!userProfile.available) return false;
    
    // Check request status and permissions
    const requestSnap = await get(ref(database, `requests/${requestId}`));
    if (!requestSnap.exists()) return false;
    const request = requestSnap.val() as BloodRequest;
    
    // Check if request is still available for acceptance
    if (request.status !== 'open' && request.status !== 'pending') return false;
    
    // Check if donor is targeted (for pending requests)
    if (request.status === 'pending' && request.requestedTo !== uid) return false;
    
    return true;
  } catch (error) {
    console.error('Error checking if request can be accepted:', error);
    return false;
  }
}

export async function getAcceptableRequests(donorUid?: string): Promise<BloodRequest[]> {
  const uid = donorUid ?? auth.currentUser?.uid;
  if (!uid) return [];
  
  try {
    // Check if donor is available
    const userSnap = await get(ref(database, `users/${uid}`));
    if (!userSnap.exists()) return [];
    const userProfile = userSnap.val();
    if (!userProfile.available) return [];
    
    // Get all requests
    const snap = await get(ref(database, 'requests'));
    if (!snap.exists()) return [];
    const all: BloodRequest[] = Object.values(snap.val() as Record<string, BloodRequest>);
    
    // Filter acceptable requests
    return all.filter((request) => {
      // Only open requests or pending requests targeted to this donor
      if (request.status === 'open') return true;
      if (request.status === 'pending' && request.requestedTo === uid) return true;
      return false;
    }).sort((a, b) => b.createdAt - a.createdAt);
  } catch (error) {
    console.error('Error getting acceptable requests:', error);
    return [];
  }
}

export async function acceptRequest(requestId: string, donorUid?: string): Promise<void> {
  const uid = donorUid ?? auth.currentUser?.uid;
  if (!uid) throw new Error('Not authenticated');
  
  // Validate that donor can accept this request
  const canAccept = await canAcceptRequest(requestId, uid);
  if (!canAccept) {
    throw new Error('You must be available to accept blood requests');
  }
  
  const snap = await get(ref(database, `requests/${requestId}`));
  if (!snap.exists()) throw new Error('Request not found');
  const current = snap.val() as BloodRequest;
  
  const isOpen = current.status === 'open';
  const isPendingTargetedToMe = current.status === 'pending' && current.requestedTo === uid;
  if (!isOpen && !isPendingTargetedToMe) {
    throw new Error('Not allowed to accept this request');
  }
  
  const now = Date.now();
  const updated: BloodRequest = {
    ...current,
    status: 'accepted',
    requestedTo: current.requestedTo ?? uid,
    acceptedAt: now,
    acceptedBy: uid,
  };
  
  await set(ref(database, `requests/${requestId}`), stripUndefined(updated as unknown as Record<string, unknown>) as unknown as BloodRequest);
  
  // Automatically create donation record when accepting request
  await createDonationRecord(requestId, uid);
}

export async function updateRequestStatus(requestId: string, status: BloodRequest['status'], donorUid?: string): Promise<void> {
  const uid = donorUid ?? auth.currentUser?.uid;
  if (!uid) throw new Error('Not authenticated');
  
  const snap = await get(ref(database, `requests/${requestId}`));
  if (!snap.exists()) throw new Error('Request not found');
  const current = snap.val() as BloodRequest;
  
  // Validate permissions based on status change
  if (status === 'accepted') {
    const canAccept = await canAcceptRequest(requestId, uid);
    if (!canAccept) {
      throw new Error('You must be available to accept blood requests');
    }
  }
  
  const now = Date.now();
  const updated: BloodRequest = {
    ...current,
    status,
    ...(status === 'accepted' && {
      requestedTo: current.requestedTo ?? uid,
      acceptedAt: now,
      acceptedBy: uid,
    }),
  };
  
  await set(ref(database, `requests/${requestId}`), stripUndefined(updated as unknown as Record<string, unknown>) as unknown as BloodRequest);
  
  // Create donation record if accepting
  if (status === 'accepted') {
    await createDonationRecord(requestId, uid);
  }
}

export async function rejectRequest(id: string): Promise<void> {
  const uid = auth.currentUser?.uid;
  if (!uid) throw new Error('Not authenticated');
  const snap = await get(ref(database, `requests/${id}`));
  if (!snap.exists()) throw new Error('Request not found');
  const current = snap.val() as BloodRequest;
  // Only targeted donor may reject targeted (pending) requests
  const isTargetedToMe = current.status === 'pending' && current.requestedTo === uid;
  if (!isTargetedToMe) {
    throw new Error('Only the targeted donor can reject this request');
  }
  const updated: BloodRequest = {
    ...current,
    status: 'rejected',
    requestedTo: current.requestedTo,
  };
  await set(ref(database, `requests/${id}`), stripUndefined(updated as unknown as Record<string, unknown>) as unknown as BloodRequest);
}

export async function cancelRequest(id: string): Promise<void> {
  const uid = auth.currentUser?.uid;
  if (!uid) throw new Error('Not authenticated');
  const snap = await get(ref(database, `requests/${id}`));
  if (!snap.exists()) throw new Error('Request not found');
  const current = snap.val() as BloodRequest;
  if (current.createdBy !== uid) {
    throw new Error('Only the creator can cancel this request');
  }
  if (current.status === 'accepted') {
    throw new Error('Cannot cancel request that has been accepted');
  }
  const updated: BloodRequest = { ...current, status: 'cancelled' };
  await set(ref(database, `requests/${id}`), stripUndefined(updated as unknown as Record<string, unknown>) as unknown as BloodRequest);
}

export async function canCancelRequest(requestId: string): Promise<boolean> {
  const uid = auth.currentUser?.uid;
  if (!uid) return false;
  const snap = await get(ref(database, `requests/${requestId}`));
  if (!snap.exists()) return false;
  const current = snap.val() as BloodRequest;
  
  // Check if user is the creator
  if (current.createdBy !== uid) return false;
  
  // Check if request is already accepted
  if (current.status === 'accepted') return false;
  
  // Check if within 10 minutes (600000 milliseconds)
  const now = Date.now();
  const tenMinutesAgo = now - (10 * 60 * 1000);
  if (current.createdAt < tenMinutesAgo) return false;
  
  return true;
}

export async function markFulfilled(id: string): Promise<void> {
  const uid = auth.currentUser?.uid;
  if (!uid) throw new Error('Not authenticated');
  const snap = await get(ref(database, `requests/${id}`));
  if (!snap.exists()) throw new Error('Request not found');
  const current = snap.val() as BloodRequest;
  if (current.createdBy !== uid) {
    throw new Error('Only the creator can mark this request as fulfilled');
  }
  const updated: BloodRequest = { ...current, status: 'fulfilled' };
  await set(ref(database, `requests/${id}`), stripUndefined(updated as unknown as Record<string, unknown>) as unknown as BloodRequest);
}

export type ListRequestsFilters = {
  status?: BloodRequest['status'] | BloodRequest['status'][];
  city?: string;
  requiredBloodGroup?: string;
  createdBy?: string;
  requestedTo?: string;
  mineOnly?: boolean; // requests I created
  toMeOnly?: boolean; // targeted to me
  openOnly?: boolean; // status === 'open'
  urgentOnly?: boolean; // urgent === true
};

export async function listRequests(filters: ListRequestsFilters = {}): Promise<BloodRequest[]> {
  // For now, fetch all and filter client-side as per Project.md guidance
  const uid = auth.currentUser?.uid ?? undefined;
  const snap = await get(query(ref(database, 'requests')));
  if (!snap.exists()) return [];
  let items: BloodRequest[] = Object.values(snap.val() as Record<string, BloodRequest>);

  if (filters.status) {
    const statuses = Array.isArray(filters.status) ? filters.status : [filters.status];
    items = items.filter((r) => statuses.includes(r.status));
  }
  if (filters.city) items = items.filter((r) => r.city === filters.city);
  if (filters.requiredBloodGroup)
    items = items.filter((r) => r.requiredBloodGroup === filters.requiredBloodGroup);
  if (filters.createdBy) items = items.filter((r) => r.createdBy === filters.createdBy);
  if (filters.requestedTo) items = items.filter((r) => r.requestedTo === filters.requestedTo);
  if (filters.mineOnly && uid) items = items.filter((r) => r.createdBy === uid);
  if (filters.toMeOnly && uid) items = items.filter((r) => r.requestedTo === uid);
  if (filters.openOnly) items = items.filter((r) => r.status === 'open');
  if (filters.urgentOnly) items = items.filter((r) => r.urgent === true);

  // Sort newest first
  items.sort((a, b) => b.createdAt - a.createdAt);
  return items;
}

export async function listDonorInbox(): Promise<{ targeted: BloodRequest[]; discoverable: BloodRequest[] }> {
  const uid = auth.currentUser?.uid;
  if (!uid) return { targeted: [], discoverable: [] };
  
  const snap = await get(query(ref(database, 'requests')));
  if (!snap.exists()) return { targeted: [], discoverable: [] };
  const all: BloodRequest[] = Object.values(snap.val() as Record<string, BloodRequest>);
  
  const targeted = all
    .filter((r) => r.requestedTo === uid && r.status === 'pending')
    .sort((a, b) => b.createdAt - a.createdAt);
  
  const discoverable = all
    .filter((r) => r.status === 'open')
    .sort((a, b) => b.createdAt - a.createdAt);
  
  return { targeted, discoverable };
}



